nuxeo-core/nuxeo-core-storage-mongodb/src/main/java/org/nuxeo/ecm/core/storage/mongodb/MongoDBRepository.java

```diff
@@ -429,15 +424,63 @@ protected void update(String op, String key, Object value) {
          * <p>
          * A conflict occurs if one key is equals to or is a prefix of the other.
          */
-        protected boolean conflicts(String key, List<String> previousKeys) {
-            String keydot = key + '.';
-            for (String prev : previousKeys) {
-                if (prev.equals(key) || prev.startsWith(keydot) || key.startsWith(prev + '.')) {
+        protected void checkForConflict(String key) {
+            List<String> pKeys = getPrefixKeys(key);
+            if (conflictKeys(key, pKeys)) {
+                newUpdate();
+            }
+            prefixKeys.addAll(pKeys);
+            keys.add(key);
+        }
+
+        protected void newUpdate() {
+            updateList.add(update = new BasicDBObject());
+            int pSize = 10;
+            int size = 10;
+            if (prefixKeys != null) {
+                pSize = prefixKeys.size();
+            }
+            if (keys != null) {
+                size = keys.size();
+            }
+            prefixKeys = new HashSet<>(pSize);
+            keys = new HashSet<>(size);
+        }
```

Comment:
`prefixKeys = new HashSet<>(prefixKeys == null ? 10 : prefixKeys.size());` etc. And please add a comment explaining why this is done instead of allocating a fixed size or clearing the previous set.